src/server/middlewares/auth-zitadel.ts

```diff
@@ -0,0 +1,54 @@
+import type {NextFunction, Request, Response} from '@gravity-ui/expresskit';
+
+import {logout} from '../controllers/zitadel';
+import {
+    generateServiceAccessUserToken,
+    introspect,
+    refreshTokens,
+    saveUserToSesson,
+} from '../utils/zitadel';
+
+export default async function (req: Request, res: Response, next: NextFunction) {
+    const {ctx} = req;
+
+    const isAuthenticated = req.isAuthenticated();
+
+    if (isAuthenticated) {
+        req.serviceUserAccessToken = await generateServiceAccessUserToken(ctx, req.user.login);
+
+        const accessToken = req.user.accessToken;
+        const refreshToken = req.user.refreshToken;
+
+        const accessTokenIntrospected = await introspect(ctx, accessToken);
+
+        if (accessTokenIntrospected) {
+            return next();
+        } else {
+            const tokens = await refreshTokens(ctx, refreshToken);
+
+            if (tokens.accessToken && tokens.refreshToken) {
+                req.user.accessToken = tokens.accessToken;
+                req.user.refreshToken = tokens.refreshToken;
+
+                await saveUserToSesson(req);
+
+                return next();
+            } else if (req.routeInfo?.ui) {
+                return logout(req, res);
```

Comment:
Missed `try/catch`? For all `req.logOut` exceptions status is `500`, or it could be meaningful?

Reply:
if you take a look at req.logOut code you will be very surprised what is inside and how they treat 'err' callback )) Anyway, it is better to brace this code with try/catch for safety